### PR TITLE
Module require resolving refactor

### PR DESCRIPTION
**Pull request by dustyo-O**
Several probs with require more than one module in different modular systems

### Changes proposed in this pull request

- If we have ym - we provide modules there, also we wrap commonJS and global deps with ym syntax
- If we do not have ym, we ignore ym deps.
- Priority: global - commonJS - ym;

### Checklist

 - [x] Documentation changed
 - [x] Tests added

### CODE REVIEW

File: lib/compiler.js [hunk 1]
```diff
@@ -155,52 +152,87 @@ Compiler.prototype.generate = function generate(code, options) {
...
+
+        'var _libs = {};',
+
+        // Provide with CommonJS
+        deps.commonJSNames.length !== 0 ?
+            deps.commonJSNames.map(function(dep) {
+              var commonPath = deps.commonJS[dep];
+              return '_libs["' + dep + '"] = ' +
+                'glob && typeof glob["' + dep + '"] ' +
+                '!== "undefined" ? ' +
```
Comment (dustyo-O) [inline lib/compiler.js]: looks like no need to do it (try to use global) here - it will get to us anyway on line 222

**Comment by miripiruni** [comment]
@dustyo-O linters are red

**Comment by dustyo-O** [comment]
fixed linters now

TODO: check that tests will fail on corrupted data (checks are checking)

File: lib/compiler.js [hunk 2]
```diff
@@ -54,28 +56,26 @@ function getDeps(requires) {
...
-    deps.ymLibs = 'engine.libs = {' + deps.ymLibs.map(function(item) {
-      return '"' + item + '":' + item;
-    }).join() + '};';
-  }
-
-  if (deps.ymVars.length) {
-    deps.ymVarsString = deps.ymVars.map(function(item) {
-      return item + ':' + item;
-    }).join('');
+    deps.ymVars = ',' + deps.ymLibs.map(function (item) {
```
Comment (dustyo-O) [inline lib/compiler.js]: we make 3 maps here, can do 1 for instead, could be faster. what do you think?

Comment (miripiruni) [reply lib/compiler.js]: @dustyo-O `generate()` is not runtime operation. No matter how it fast or slow we use it to generate bundles for browser.

File: test/requires-test.js [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+var FAKE_COMMON_2MODULES = 'var require=(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module \'"+o+"\'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({"fake":[function(require,module,exports){ module.exports = { getText: function () { return "Hello templates!"; } }; },{}], "fake1":[function(require,module,exports){ module.exports = { getText: function () { return "Good-bye templates...."; } }; },{}]},{},[]);';
+// test texts
+var TEXT = 'Hello templates!';
+var TEXT1 = 'Good-bye templates....';
+/* jscs:enable */
+var chai = require('chai');
+var chaiAsPromised = require('chai-as-promised');
+chai.use(chaiAsPromised);
+
+describe('Require modules', function () {
```
Comment (miripiruni) [inline test/requires-test.js]: Needed to move back tests to `test/api-generate-test.js`.

File: test/requires-test.js [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+        var bundle = bemhtml.generate('', {
+          requires: { textModule: { globals: 'text' } }
+        });
+
+        var sandbox = { global: {}, exports: {} };
+        sandbox.module = { exports: sandbox.exports };
+
+        vm.runInNewContext(code + EOL + bundle, sandbox);
+
+        var result = sandbox.exports.bemhtml.compile(function () {
```
Comment (miripiruni) [inline test/requires-test.js]: Code style: notice that `function() {` in other files in the repo doesn’t have space before `()` brackets.

File: lib/compiler.js [hunk 5]
```diff
@@ -54,28 +56,26 @@ function getDeps(requires) {
...
+      return item;
+    }).join(',');
```
Comment (miripiruni) [inline lib/compiler.js]: What?

```js
deps.ymLibs.map(function(item) { return item; }).join(',');
```

`map` is useless here.

File: lib/compiler.js [hunk 6]
```diff
@@ -155,52 +155,89 @@ Compiler.prototype.generate = function generate(code, options) {
...
+          return 'modules.define(' +
+            '"' + dep + '"' + ',' +
+            'function (provide, prev) {' +
+            'provide(glob["' + globName + '"] || prev);' +
+            '});';
+        }).join(EOL) : '',
+
+        'modules.define("' + exportName + '",' +
+          JSON.stringify(deps.ymLibs) + ',' +
+          'function(provide' + deps.ymVars + ') { ' +
```
Comment (miripiruni) [inline lib/compiler.js]: let’s place `deps.ymVars` inline?

```js
'provide, ' + deps. … .join(', ')
```

File: lib/compiler.js [hunk 7]
```diff
@@ -28,10 +28,12 @@ function getDeps(requires) {
     globalNames: [],
     commonJS: {},
     commonJSNames: [],
-    ym: [],
-    ymVars: [],
-    ymVarsString: '',
+    ym: {},
+    ymNames: [],
     ymLibs: [],
+    ymVars: '',
```
Comment (miripiruni) [inline lib/compiler.js]: Why we need both `ymVars` and `ymNames`?
Previuosly, I used only `ymVars: []` and  `ymVars.join(', ')`

So many variables… I’m try to reduce their number.

**Comment by miripiruni** [comment]
JFYI: we just discuss details with @dustyo-O. Decided to finish PR this evening together.

**Comment by dustyo-O** [comment]
ok guys, i've sawed the whole diff into step-by-step commits. i hope, this helps - happy reading!

into the f98319f i've added some alpha-version documentation, please read carefully and feel free to suggest changes. I'll add english translation before merge.

File: docs/ru/3-api.md [hunk 8]
```diff
@@ -486,6 +486,18 @@ block('button').content()(function () {
...
+1) global
+2) CommonJS
+3) YModules (при условии, что модульная система доступна)
+Если на любом из шагов соответствующий модуль был найден, то дальнейшие шаги игнорируются - в шаблоны попадает первый найденный.
+
+Таким образом, если модуль доступен в глобальной переменной, то внутрь шаблонов попадет значение переменной, даже если существует нужный модуль в CommonJS/YM.
+
+И также, модуль из CommonJS приоритетнее модуля в YModules.
+
+Это позволяет, например, переопределять ваши модули на статические объекты в целях тестирования в соответствующих конфигурациях.
```
Comment (miripiruni) [inline docs/ru/3-api.md]: Кажется, что эту фразу можно пропустить.

File: lib/compiler.js [hunk 9]
```diff
@@ -155,51 +151,93 @@ Compiler.prototype.generate = function generate(code, options) {
...
-     'if (typeof modules === "object") {',
-       'modules.define("' + exportName + '",' +
-         JSON.stringify(deps.ym) + ',' +
-         'function(provide' + deps.ymVars + ') { ' +
-            'var engine = buildBemXjst({' + deps.ymVarsString + '});' +
-            deps.ymLibs +
+      'if (typeof modules === "object") {',
+
+      // re-define YM modules to require keys
+      deps.ymNames.length !== 0 ?
```
Comment (miripiruni) [inline lib/compiler.js]: Можно оставить просто `deps.ymNames.length`…

**Pull request merged**